components/pixels/snapshot.rs

```diff
@@ -81,34 +82,50 @@ impl SnapshotAlphaMode {
     }
 }
 
-#[derive(Clone, Debug, Deserialize, MallocSizeOf, Serialize)]
+/// The data in a [`Snapshot`]. If created via shared memory, this will be
+/// the `SharedMemory` variant, but otherwise it is the `Owned` variant.
+/// any attempt to mutate the [`Snapshot`] will convert it to the `Owned`
+/// variant.
+#[derive(Clone, Debug, MallocSizeOf)]
 pub enum SnapshotData {
-    // TODO: https://github.com/servo/servo/issues/36594
-    // IPC(IpcSharedMemory),
+    SharedMemory(
+        #[conditional_malloc_size_of] Arc<IpcSharedMemory>,
+        Range<usize>,
+    ),
     Owned(Vec<u8>),
 }
 
-impl Deref for SnapshotData {
-    type Target = [u8];
-
-    fn deref(&self) -> &Self::Target {
+impl SnapshotData {
+    fn to_vec(&self) -> Vec<u8> {
         match &self {
-            // Data::IPC(ipc_shared_memory) => ipc_shared_memory,
-            SnapshotData::Owned(items) => items,
+            SnapshotData::SharedMemory(data, byte_range) => Vec::from(&data[byte_range.clone()]),
+            SnapshotData::Owned(data) => data.clone(),
         }
     }
 }
 
 impl DerefMut for SnapshotData {
     fn deref_mut(&mut self) -> &mut Self::Target {
         match self {
-            // Data::IPC(ipc_shared_memory) => unsafe { ipc_shared_memory.deref_mut() },
+            SnapshotData::SharedMemory(..) => {
+                *self = SnapshotData::Owned(self.to_vec());
+                &mut *self
+            },
```

Comment:
This is the magic that converts a shared memory `SnapshotData` into a mutable one when an attempt to mutate the data is made.